File: python/p3.py
import math

def gen(n):
    yield 2
    yield 3
    max = int(math.sqrt(n))
    for i in xrange(1, int((n + 1) / 6)):
        for rslt in (6*i - 1, 6*i + 1):
            if rslt > max:
                return
            yield rslt

def find_largest_prime_factor(num):
    rslt = n = num

    g = iter(gen(num))

    i = next(g)
    while n > 1:
        while n % i == 0:
            n /= i
            rslt = i
        i = next(g)

    return rslt

print find_largest_prime_factor(600851475143)

File: lua/p3.lua
function gen(n)
    return coroutine.create(function ()
        coroutine.yield(2)
        coroutine.yield(3)
        local max = math.sqrt(n)
        for i = 1, (n + 1) / 6 do
            for x, rslt in ipairs({6*i - 1, 6*i + 1}) do
                if rslt > max then
                    return 0
                end
                coroutine.yield(rslt)
            end
        end
    end)
end

function find_largest_prime_factor(num)
    local rslt = num
    local n = num
    local co = gen(num)
    local stat, i
    stat, i = coroutine.resume(co)

    while n > 1 do
        while n % i == 0 do
            n = n / i
            rslt = i
        end
        stat, i = coroutine.resume(co)
        if not stat or i == 0 then
            break
        end
    end

    return rslt
end

print(find_largest_prime_factor(600851475143))
